wen-restart/src/wen_restart.rs

```diff
@@ -32,17 +32,42 @@ use {
         thread::sleep,
         time::Duration,
     },
+    thiserror::Error,
 };
 
 // If >42% of the validators have this block, repair this block locally.
 const REPAIR_THRESHOLD: f64 = 0.42;
 
+#[derive(Error, Debug)]
+pub enum WenRestartError {
+    #[error("Protobuf Decode error: {0}")]
+    DecodeError(#[from] prost::DecodeError),
+    #[error("Protobuf Encode error: {0}")]
+    EncodeError(#[from] prost::EncodeError),
+    #[error("Exiting")]
+    Exiting,
+    #[error("Invalid Last Vote Type: {0:?}")]
+    InvalidLastVoteType(VoteTransaction),
+    #[error("File IO error: {0}")]
+    IoError(#[from] std::io::Error),
+    #[error("Missing last voted fork slots")]
+    MissingLastVotedForkSlots,
+    #[error("Hash parse error: {0}")]
+    ParseHashError(#[from] solana_sdk::hash::ParseHashError),
+    #[error("Pubkey parse error: {0}")]
+    PubkeyParseError(#[from] solana_sdk::pubkey::ParsePubkeyError),
+    #[error("Gossip push RestartLastVotedForkSlotsError: {0}")]
+    RestartLastVotedForkSlotsError(#[from] RestartLastVotedForkSlotsError),
+    #[error("Unexpected state: {0:?}")]
+    UnexpectedState(wen_restart_proto::State),
```

Comment:
I think it would be better to organize these by state transition:

```
pub enum SendLastVotedForkError {

}

pub enum LastVotedForkSlotsError {

}
```

1. Then `wait_for_wen_restart()` still returns `Box<dyn std::error::Error>`
2. For testing for each state transition, we make sure all of their particular errors write properly to the record and we can restart

Reply:
Hmm, when you say "write properly to the record", do you mean writing to the logs or to the protobuf as well? Since some of the errors might be transient, I didn't plan to write the errors to the protobuf.

Reply:
Oh I didn't mean write the error out. I mean the `progress` data structure you write is what you expect after the error happens

Reply:
I did the change but it actually felt wrong IMHO, several reasons:
1. complicating public API for the sake of testing
pub enum xxxError is public interface, if you think about how it's used, all people see from outside the module is wait_for_wen_restart, so it feels natural to have WenRestartError returned, it feels unnatural to have several Error enum returned, leaking our implementation details (which they shouldn't care about).
2. harder to write tests
I could of course test each function individually, but I do have to test they work in wait_for_wen_restart. If I only need to test is_err() then it's all fine, but if I want to test for specific error (e.g. file permission wrong), I had to do a downcast on dyn error, I had to constantly look back to error definitions to see what type of error I'm now expecting
3. common functions harder to implement and change
The fact that we do need to write the protobuf means I need to call write_wen_restart() from several places, but they will return different xxxError types, so I need to create yet another WriteWenRestartError and convert to other Error types, this is code duplication and unnecessary conversion. If we move part of the function to another internal method, we need to create yet another Error struct?

All in all, this feels wrong, maybe I'm misunderstanding what you described.

Reply:
>pub enum xxxError is public interface, if you think about how it's used, all people see from outside the module is wait_for_wen_restart, so it feels natural to have WenRestartError returned, it feels unnatural to have several Error enum returned, leaking our implementation details (which they shouldn't care about).

Here we're not writing a standalone library so I think a boxed error is fine. The purpose of this error is to make attribution of errors/debugging easier, which categorizing the errors does. We can now trace exactly which step in the restart went wrong. I think we could also introduce a top level `WenRestartError` that wraps the errors like so:

```
enum WenRestartError {
   SendLastVotedForkError(SendLastVotedForkError),
   AggregateLastVotedForkSlotsError(AggregateLastVotedForkSlotsError),
  ...
}
```

>I could of course test each function individually, but I do have to test they work in wait_for_wen_restart. If I only need to test is_err() then it's all fine, but if I want to test for specific error (e.g. file permission wrong), I had to do a downcast on dyn error, I had to constantly look back to error definitions to see what type of error I'm now expecting

Hmm I don't think you have to test every error case in `wait_for_wen_restart()`.  I think if the state transition functions are `a-> b -> c`, then:

1. You just have to test that each transition `a`, `b`, and `c` individually will run successfully to "completion" across all cases. "Completion" in this case means the ending output passes the verification for the initial input of the next state transition. Concretely this means testing in the happy case the state is correct, and that on all declared errors, they can recover and then run to completion. 
2. From 1, inductively, you can then conclude that a -> b -> c will run correctly.
3. All you need to test then in `wait_for_wen_restart()` is that the order of calls `a -> b -> c` is correct and the arguments are being passed correctly. Concretely this means just testing one happy path.

Reply:
But we do have write_wen_restart_proto in a few of the steps, that means I have to replicate its error in all those steps? This seems like unnecessary repetition.
Also, I understand how splitting them into several errors make testing easier to read now, but it also makes changes in the future painful.
Let's say in the future for whatever reason we want to reorganize SendLastVotedForkSlots and AggregateLastVotedForkSlots into three steps, ForkSlotsStage1, ForkSlotsStage2, ForkSlotsStage3, now you need to change the Error enums everywhere as well.
Splitting it into the current steps is a specific implementation detail in the current design, in the future we might ditch the current design and do something else, the public interface shouldn't have to change at all. The public interface only cares that wait_on_wen_restart succeeded, or failed on IO, or failed on Gossip, or failed on bad input, it shouldn't care which step the failure is in, so the caller shouldn't need to match the specific suberror inside WenRestartError.

Reply:
Thinking about it more, I think I agree that failures should be tested on individual functions rather than wait_on_wen_restart(), so I changed accordingly.

The only thing I'm hesitating is that I still don't see how splitting WenRestartError into individual errors helps us. If the proto file is suddenly not writable, we need to solve that problem no matter what stage WenRestart is in, so I don't think replicating file IO errors everywhere really helps the user, it just complicates the public API.

Reply:
How about this, we guarantee no errors on the state transitions 😄 . I.e. we guarantee the state going in and the state coming out of the records are always valid.

1. All of the current errors are verification errors on the initial state: https://github.com/solana-labs/solana/blob/645452f52a8103dd283f36b855977e07025112db/wen-restart/src/wen_restart.rs#L88-L92 and 
https://github.com/solana-labs/solana/pull/33892/files#diff-c32a5c15bdd23820489e6f8af8b305aee344228d0cb6789e554a988d57d54bd7R70-R71. These validations can be done in the beginning on entry into the state machine in `wait_for_wen_restart`, as a `validate_state_1()`, `validate_state_2()`, etc. These can then be unit tested. Ideally I think we would change `WenRestartProgress` to be an enum with a separate state for each step in this state machine, i.e. 

```
enum WenRestartProgress {
       Init,
       SendForkSlots(SendForkSlots),
       AggregateForkslots(AggregateForkSlots),
}
```

and then at the end of each state transition we just spit out the next iteration of the enum

2. Then the only acceptable error is logging errors which we can then enforce is the only acceptable error. 
3. Then if we test that `write_wen_restart_records()` either writes a whole valid record OR doesn't write a record/i.e. errors, then we know the state must always be valid.

Reply:
I can move the initialization errors into initialize(), but I can't guarantee there are no other errors after initialize(). For example, push_restart_last_voted_fork_slots() could return error. And we might want to throw out error in the future while running.

Reply:
you're right on this 😃 

If we don't want a separate enum per state transition, then at minimum I would like to be able to print out the backtrace of where that error occurred. Maybe this: https://stackoverflow.com/questions/75060346/how-can-i-get-file-and-line-number-where-an-error-was-returned

Reply:
Changed to anyhow instead of thiserror for backtrace, also moved initialization errors into initialize().